extensions/common/xwalk_external_instance.cc

```diff
@@ -34,18 +34,25 @@ XWalkExternalInstance::~XWalkExternalInstance() {
 
 void XWalkExternalInstance::HandleMessage(scoped_ptr<base::Value> msg) {
   XW_HandleMessageCallback callback = extension_->handle_msg_callback_;
-  if (!callback) {
+  XW_HandleBinaryMessageCallback binary_callback =
+      extension_->handle_binary_msg_callback_;
+  if (!callback && !binary_callback) {
     LOG(WARNING) << "Ignoring message sent for external extension '"
                  << extension_->name() << "' which doesn't support it.";
     return;
   }
 
   std::string string_msg;
-  if (!msg->GetAsString(&string_msg)) {
+  const base::BinaryValue* binary_msg = nullptr;
+  if (callback && msg->GetAsString(&string_msg)) {
+    callback(xw_instance_, string_msg.c_str());
+  } else if (binary_callback && msg->GetAsBinary(&binary_msg)) {
+    binary_callback(xw_instance_, binary_msg->GetBuffer(),
+                    binary_msg->GetSize());
```

Comment:
If this is by design, it's better to raise an assertion/exception while extensions with string message callback try to register a binary message callback.